src/coronawhy_vt/utils/local_processing.py

```diff
@@ -0,0 +1,375 @@
+
+    :param paper (json): The paper to be parsed
+
+    :return: Lines of text in the paper title
+    """
+    return [{'text': paper['metadata']['title']}]
+
+
+def abstract_blocks(paper: dict) -> list:
+    """Retrieve abstract block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper abstract
+    """
+    return paper['abstract']
+
+
+def body_blocks(paper: dict) -> list:
+    """Retrieve body block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper body
+    """
+    return paper['body_text']
+
+
+# TODO the reference nlp is an external dependence, either add into function or make it a class
+nlp = spacy.load('en_core_web_lg')
+
+
+def pull_mentions_lemmatized(
+        paths: list,
+        block_selector: Callable[[dict], list],
+        sec_name: str,
+        words: list,
+        replace_dict: dict = None) -> dict:
+    """Aggregate the positional features of the lemmatized text corpus.
+
+    :param paths: pathlib.Path objects corresponding to directories containing
+        corpus to aggregate features from.
+    :param block_selector: The function to retrieve the relevant block from text
+    :param sec_name: Corresponds with block_selector to select the relevant
+        block of text. Options are 'title', 'abstract', 'body'
+        TODO seems like this should be pulled into the function as it depends
+        entirely on the block_selector argument
+    :param words: List of lemmatized words
+    :param replace_dict: Takes a dict of {token: lemmatized underscore token}
+        and replaces occurences of token in text with its value
+
+    :return: {'identifier feature': list of occurrences}
+    """
+    positions = []
+    found_words = []
+    section = []
+    block_id = []
+    block_text = []
+    paper_id = []
+
+    tokenized_words = []
+    for w in words:
+        tokenized_words.append(nlp(w.lower())[0].lemma_)
+    for path in paths:
+        files = os.listdir(path)
+        for p in files:
+
+            readfile = open(path / p, 'r')
+            paper = json.load(readfile)
+            blocks = block_selector(paper)
+
+            for b in range(0, len(blocks)):
+                text = blocks[b]['text'].lower()
+                if(replace_dict is not None):
+                    text = run_replace(text, replace_dict)
+                text = nlp(text)
+                for t in text:
+                    for w in tokenized_words:
+                        if(w == t.lemma_):
+                            section.append(sec_name)
+                            found_words.append(w)
+                            positions.append(t.idx)
+                            block_text.append(blocks[b]['text'])
+                            block_id.append(b)
+                            paper_id.append(p[:-5])
+    # TODO discuss pros/cons of this format vs a list of tuples
```

Comment:
I like dicts better! Much more self-documentation.

Reply:
I agree, however in this case the dict values represent individual lists which then have to be aggregated to get entries for similarities. i.e. 

`mentions = pull_mentions_lemmatized()`
`first_entry = (mentions['sha'][0], mentions['blockid'][0], mentions['word'][0], mentions['sec'][0], mentions['pos'][0], mentions['block'][0])`

whereas if we used a list of tuples (or even a list of dicts)

`mentions = pull_mentions_lemmatized`
`first_entry = mentions[0]`

Seems like it would be cleaner to use for the end user.

Reply:
Maybe a namedtuple would be best then if the list is always the same size